### PR TITLE
[viomi branch] refactor to ES6 classes

### PR DESCRIPTION
**Pull request by alexkn**
roborock/viomi classes as mentioned in #433

I tested only the roborock part.
@rumpeltux please test the viomi part.

### CODE REVIEW

File: lib/devices/Viomi.js [hunk 1]
```diff
@@ -80,85 +162,4 @@ const ERROR_MAP = Object.freeze({
...
-    });
-};
-
-Viomi.prototype.getCurrentStatus = function() {
-    return Promise.resolve(this.status);
-};
-
-Viomi.prototype.pollMap = function() {
-    clearTimeout(this.pollMapTask);
-    this.sendLocalWithPromise("set_uploadmap", [2], {timeout: 2000}).finally(() => {
```
Comment (rumpeltux) [inline lib/devices/Viomi.js]: Hmm sendLocalWithPromise should have these options too, but then we also need to support this for the cloudSocket… I can do that another time though.

File: lib/miio/Status.js [hunk 2]
```diff
@@ -4,16 +4,17 @@
  */
 const Status = function() {
     /** List of allowed values as per Roborock.STATUS_MAP */
-    this.state = "UNKNOWN";
+    this.state = 0;
```
Comment (rumpeltux) [inline lib/miio/Status.js]: I think using strings or an enum here is strictly better. And roborock updateStatus also sets this to a string.

File: lib/miio/Status.js [hunk 3]
```diff
@@ -4,16 +4,17 @@
...
     this.error_code = null;
     this.battery = null;
     this.clean_time = null;   // Seconds?
     this.clean_area = null;   // mm²
     this.error_code = null;   // TODO
     this.map_present = false; // TODO
-    this.in_cleaning = false; // TODO
+    this.in_cleaning = 0; // TODO
```
Comment (rumpeltux) [inline lib/miio/Status.js]: Why int?

File: lib/miio/Status.js [hunk 4]
```diff
@@ -4,16 +4,17 @@
...
     this.fan_power = null;
     this.dnd_enabled = false;
+    this.lab_status = 0;
```
Comment (rumpeltux) [inline lib/miio/Status.js]: If adding a new field here, you should precisely document what it means.

File: lib/miio/Status.js (hunk 2 above)
Comment (alexkn) [reply lib/miio/Status.js]: That was only a fix due to type checking.
There is Code which uses it as number.

File: lib/webserver/WebServer.js [hunk 5]
```diff
@@ -563,7 +563,7 @@ const WebServer = function (options) {
                 ...req.body.virtual_walls.map(wall => [1, ...wall])
             ];
 
-            if (self.vacuum.model === "roborock.vacuum.s5") { //TODO: Move these magic strings to somewhere else
+            if (self.model.RoborockApi) {
```
Comment (Hypfer) [inline lib/webserver/WebServer.js]: Finally

File: lib/miio/Status.js (hunk 3 above)
Comment (alexkn) [reply lib/miio/Status.js]: That was only a fix due to type checking.
There is Code which uses it as number.

File: lib/miio/Status.js (hunk 2 above)
Comment (rumpeltux) [reply lib/miio/Status.js]: The one in webserver.js right? Should fix those locations instead.

File: lib/miio/Status.js (hunk 4 above)
Comment (alexkn) [reply lib/miio/Status.js]: That was also a fix due to type checking.
There is Code which uses it as number, but I don't know what each number means.

File: lib/devices/MiioVacuum.js [hunk 6]
```diff
@@ -9,169 +9,177 @@ const Status = require("../miio/Status");
...
-        doTimesync: false
-    });
-}
-
-MiioVacuum.FAST_MAP_PULLING_STATES = [
-    "REMOTE_CONTROL_ACTIVE", "CLEANING", "RETURNING_HOME", "MANUAL_MODE", "SPOT_CLEANING",
-    "DOCKING", "GOING_TO_TARGET", "ZONED_CLEANING"
-];
+class MiioVacuum {
+    /* @param options {object}
```
[the remark this answers is not in the document]
Comment (alexkn) [reply lib/devices/MiioVacuum.js]: Yes, I missed it, either you fix it yourself or I am doing it this evening.

File: lib/miio/Status.js (hunk 4 above)
Comment (rimrul) [reply lib/miio/Status.js]: Looking at [client/settings-persistent-data.html#L102-L167](https://github.com/Hypfer/Valetudo/blob/master/client/settings-persistent-data.html), it seems like `1` means `persistent data enabled` and every other value means `persistent data disabled`.

File: lib/miio/Model.js [hunk 7]
```diff
@@ -1,17 +1,20 @@
...
-    this.name = name;
-};
+class Model {
+    /**
+     * @param {string} name
+     * @constructor
+     */
+    constructor(name) {
+        this.viomiApi = name == "viomi.vacuum.v7";
+        this.RoborockApi = name == "roborock.vacuum.s5";
```
Comment (rimrul) [inline lib/miio/Model.js]: This works for what you use it for, but doesn't really indicate what the naming implies. The naming implies that for a Roborock vacuum the property `RoborockApi` is true, while for a Viomi vacuum the property `viomiApi` is true. Instead the property `viomiApi` is true for all currently supported Viomi vacuums and false for all currently supported Roborock vacuums. The `RoborockApi` property does not properly indicate that the `Roborock` vacuum API should be used, but that the robot is a Gen 2 Roborock vacuum and supports persistent data. 

Thinking about future supported devices, maybe we shouldn't set booleans here to define what implementation of `MiioVacuum` to use, but rather set a single property that already references the correct implementation to use.

Comment (alexkn) [reply lib/miio/Model.js]: This should be refactored later, it's not the topic of this pull request.

Comment (rimrul) [reply lib/miio/Model.js]: Fair enough.

File: lib/miio/Status.js (hunk 2 above)
Comment (alexkn) [reply lib/miio/Status.js]: It is used almost everywhere as a number, even your fakerobot.js sets it to 2.
Maybe it should be refactored later, it's not the topic of this pull request.

File: lib/miio/Status.js (hunk 3 above)
Comment (alexkn) [reply lib/miio/Status.js]: It is used everywhere as a number.
Maybe it should be refactored later, it's not the topic of this pull request.

**Comment by alexkn** [comment]
In my opinion, this can be merged.

File: lib/miio/Status.js (hunk 2 above)
Comment (rumpeltux) [reply lib/miio/Status.js]: fakerobot sends a json object not a Status one.
It is converted by updateStatus to a string: https://github.com/Hypfer/Valetudo/blob/viomi/lib/devices/Roborock.js#L173
The remaining usage in webserver.js as an int is a bug that should be fixed.
But not necessarily here. Feel free to merge and we can fix it later.

**Pull request merged**